File: Códigos/C_Codes/Lista Circular Encadeada/circularLinkedList.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <locale.h>

typedef struct Node {
    int data;
    struct Node* next;
} Node;

Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = newNode; 
    return newNode;
}

void insertAtHead(Node** head, int data) {
    Node* newNode = createNode(data);

    if (*head == NULL) {
        *head = newNode;
    } else {
        Node* tail = (*head)->next;

        newNode->next = *head;
        while (tail->next != *head) {
            tail = tail->next;
        }
        tail->next = newNode;
        *head = newNode;
    }
}

void removeNode(Node** head, int data) {
    if (*head == NULL) {
        printf("A lista está vazia.\n");
        return;
    }

    Node* current = *head;
    Node* prev = NULL;

    do {
        if (current->data == data) {
            if (prev == NULL) {
                Node* tail = (*head)->next;
                while (tail->next != *head) {
                    tail = tail->next;
                }
                if (current->next == *head) {
                    *head = (current->next != current) ? current->next : NULL;
                }
                tail->next = *head;
            } else {
                prev->next = current->next;
            }
            free(current);
            printf("Elemento %d removido.\n", data);
            return;
        }
        prev = current;
        current = current->next;
    } while (current != *head);

    printf("Elemento %d não encontrado.\n", data);
}

void printList(Node* head) {
    if (head == NULL) {
        printf("A lista está vazia.\n");
        return;
    }

    Node* current = head;
    do {
        printf("%d -> ", current->data);
        current = current->next;
    } while (current != head);
    printf("(head)\n");
}

int main() {
    setlocale(LC_ALL, "Portuguese");
    Node* head = NULL;
    int choice, data;

    srand(time(NULL));

    while (1) {
        if(head != NULL){
            printf("Lista: ");
            printList(head);
            printf("\n1. Inserir na lista\n2. Remover da lista\n3. Sair\nEscolha uma opção: ");
            scanf("%d", &choice);
        } else {
            printf("\n1. Inserir na lista\n3. Sair\nEscolha uma opção: ");
            scanf("%d", &choice);
            if (choice == 2) {
                printf("Opção Inválida\n");
                continue;
            } 
        }

        switch (choice) {
            case 1:
                data = rand() % 100;
                insertAtHead(&head, data);
                printf("Número %d inserido na lista.\n", data);
                break;
            case 2:
                if (head == NULL) {
                    printf("A lista está vazia, não é possível remover elementos.\n");
                    break;
                }
                printf("Digite o número a ser removido: ");
                scanf("%d", &data);
                removeNode(&head, data);
                break;
            case 3:
                if (head != NULL) {
                    Node* current = head;
                    Node* nextNode;
                    do {
                        nextNode = current->next;
                        free(current);
                        current = nextNode;
                    } while (current != head);
                }
                return 0;
            default:
                printf("Opção inválida.\n");
        }
    }
}

File: Códigos/Python Codes/Lista Circular Encadeada/circular_linked_list.py
import random


class Node:
    def __init__(self, data):
        self.data = data
        self.next = self  # Circular reference


def create_node(data):
    return Node(data)


def insert_at_head(head, data):
    new_node = create_node(data)
    if head is None:
        head = new_node
    else:
        tail = head
        while tail.next != head:
            tail = tail.next
        new_node.next = head
        tail.next = new_node
        head = new_node
    return head


def remove_node(head, data):
    if head is None:
        print("A lista está vazia.")
        return head

    current = head
    prev = None

    while True:
        if current.data == data:
            if prev is None:  # Removing head
                head = remove_head(head)
            else:
                prev.next = current.next
            print(f"Elemento {data} removido.")
            return head
        prev = current
        current = current.next
        if current == head:
            break

    print(f"Elemento {data} não encontrado.")
    return head


def remove_head(head):
    if head is None:
        return None

    tail = head
    while tail.next != head:
        tail = tail.next

    if head.next == head:  # Only one node in the list
        return None

    new_head = head.next
    tail.next = new_head
    return new_head


def is_list_empty(head):
    return head is None


def print_list(head):
    if is_list_empty(head):
        print("A lista está vazia.")
        return

    current = head
    while True:
        print(f"{current.data} -> ", end='')
        current = current.next
        if current == head:
            break
    print("(head)")


def print_menu_and_list(head):
    print_menu()
    if head is not None:
        print("Lista: ", end='')
        print_list(head)


def print_menu():
    print("\n1. Inserir na lista\n2. Remover da lista\n3. Sair\nEscolha uma opção: ", end='')


def generate_random_data(min_value=0, max_value=99):
    return random.randint(min_value, max_value)


def handle_insert(head):
    data = generate_random_data()
    head = insert_at_head(head, data)
    print(f"Número {data} inserido na lista.")
    return head


def handle_remove(head):
    if head is None:
        print("A lista está vazia, não é possível remover elementos.")
    else:
        data = int(input("Digite o número a ser removido: "))
        head = remove_node(head, data)
    return head


def handle_exit(head):
    while head is not None:
        next_node = head.next
        del head
        head = next_node


def main():
    head = None
    while True:
        print_menu_and_list(head)
        choice = int(input())

        if choice == 1:
            head = handle_insert(head)
        elif choice == 2:
            head = handle_remove(head)
        elif choice == 3:
            handle_exit(head)
            break
        else:
            print("Opção inválida.")


if __name__ == "__main__":
    main()
